enterprise/backend/src/metabase_enterprise/gsheets/api.clj

```diff
@@ -0,0 +1,334 @@
+;; the folder and starts syncing the data to Metabase. This data is then available to the user in Metabase as tables.
+;;
+;; ## Steps to connect a new google drive folder
+;;
+;; - An admin clicks Add Data > Connect Google Sheets
+;; - FE sends :get "api/ee/gsheets/service-account" to get service-account email
+;;   - If the FE cannot find the service-account email, it does not show the "connect to google sheets" (ctgs) button.
+;; - user clicks ctgs button -> FE shows instructions for how to connect a folder:
+;;   - user shares their Google Drive folder w/ service-account email
+;;   - user copies their Google Drive share link into MB form and hits submit
+;; - FE sends :post "api/ee/gsheets/folder" w/ folder url string as `request.body.url`
+;;   - BE forwards request to :post "/api/v2/mb/connections" w/ body:
+;;     `{:type "gdrive" :secret {:resources ["the-url"]}}`
+;;     - on unexceptional status:
+;;       - BE sets `gsheets.status` to `"loading"`
+;;       - BE returns the gsheets shape: `{:status "loading" :folder_url "the-url" :folder-upload-time <epoch-time> :gdrive/conn-id <uuid>}}`
+;;     - on exceptional status, BE returns a message like: "Unable to setup drive folder sync. Please check that the
+;;       folder is shared with the proper service account email and sharing permissions."
+;;
+;; ## Polling
+;; - FE polls :get "api/ee/gsheets/folder" until `body.status` == `complete`
+;; - BE forwards requests to :get "/api/v2/mb/connection/<gdrive-conn-id>", filtering for the google drive connection
+;;   - If the connection doesn't exist: `{:error "google drive connection not found."}`
+;;   - If the connection exists and is NOT active, return loading state in the response
+;;   - If the connection exists and is active AND the latest sync of the attached datawarehouse is AFTER the sync time
+;;         on Harbormaster's gdrive connection, return `body.status` == "complete".
+;; - FE sees status == "complete"
+;;
+;; ### What does "complete" mean?
+;; When the status is "complete", the user can access their google sheets info inside Metabase.
+;; This means that both:
+;; 1. the gdrive connection has been marked active
+;; 2. the attached datawarehouse has completed syncing inside metabase
+;; To check this, we compare the last sync time of the attached datawarehouse with the last sync time of the gdrive.
+;;
+;; ### What if the sync takes too long or never ends?
+;; Upon the successful creation of a gdrive connection, we start a timer that will error out if the sync does not
+;; complete within [[*folder-setup-timeout-seconds*]] seconds. If the sync does not complete within this window, we
+;; reset the gsheets status to `{:status "not-connected"}` and return an error for the FE to show.
+;;
+;; ## Steps to disconnect a google drive folder
+;; - FE sends request to :delete "/folder"
+;;   - MB loops over all gdrive connection ids and deletes them
```

Comment:
You'd think there could only 1, but it's possible for there to be more in rare cases.